// rest-helper/lib/main.js
var path = require('path');
var fs = require('fs');

var RESTHelper = function (datapath) {
  this.datapath = datapath;
};

// req and res are the http request and response respectively
RESTHelper.prototype.findAll = function (req, res) {
  var self = this;

  loadDir(self.datapath, null, function (err, data) {
    if (err) {
      console.log(err);
      res.send(404);
    } else {
      res.send(200, data);
    }
  });        
};

RESTHelper.prototype.findById = function (req, res) {
  var id = req.params.id;
  var self = this;

  fs.readFile(path.join(self.datapath, id+".json"), function(err, data) {
    if (err) {
      console.log(err);
      res.send(204);
    } else {
      res.send(200, data);
    }
  }); //  fs.readFile
};

RESTHelper.prototype.add = function (req, res) {
  var self = this;
  var strData = '';
  
  if (typeof req.body !== 'undefined') {
    strData = JSON.stringify(req.body);
    if (strData != '') {
      var filepath = self.datapath;       
      fs.readdir(filepath, function(err, files) {
        if (err) {
          fs.mkdir(filepath, function(err) {
            req.body.id = 1;
            saveFile(filepath, id, strData, 'json', function(err, result) {
              if (err) {
                console.log(err);
                res.send(500);
              } else {
                res.send(200, result);
              }
            });
          });
        } else {
          // set id = max. of existing files
          getMaxFilename(files, function(err, id) {
            req.body.id = id;
            strData = JSON.stringify(req.body);
            saveFile(filepath, id, strData, 'json', function(err, result) {
              if (err) {
                console.log(err);
                res.send(500);
              } else {
                res.send(200, result);
              }
            });              
          });

        };
      });   
    
    } else {
      console.log("empty request body");
      res.send(500);
    }        
  } else {
    console.log("undefined request body");
    res.send(500);
  }
};

RESTHelper.prototype.updateById = function (req, res) {
  var id = req.params.id;
  var self = this;
  var strData = '';
  
  if (typeof req.body !== 'undefined') {
    strData = JSON.stringify(req.body);
    if (strData != '') {
      var filepath = self.datapath;
      fs.exists(filepath, function(exists) {
        if (exists) {
            saveFile(filepath, id, strData, 'json', function(err, result) {
            if (err) {
              console.log(err);
              res.send(500);
            } else {
              res.send(200, result);
            }
          });
        } else {
          fs.mkdir(filepath, function(err) {
            saveFile(filepath, id, strData, 'json', function(err, result) {
              if (err) {
                console.log(err);
                res.send(500);
              } else {
                res.send(200, result);
              }
            });
          });
        }
      });
    } else {
      console.log("empty request body");
      res.send(500);
    }    
  } else {
    console.log("undefined request body");
    res.send(500);
  }
};

RESTHelper.prototype.removeAll = function (req, res) {
  var id = req.params.id;
  var self = this;

  removeDir(self.datapath, null, function (err) {
    if (err) {
      console.log(err);
      res.send(500);
    } else {
      res.send(200);
    }
  });   
};

RESTHelper.prototype.removeById = function (req, res) {
  var id = req.params.id;
  var self = this;

  fs.readFile(path.join(self.datapath, id+".json"), function(err, data) {
    if (err) {
      console.log(err);
      res.send(404);
    } else {
      fs.unlink(path.join(self.datapath, id+".json"), function(err) {
        if (err) {
          console.log(err);
          res.send(500);
        } else {
          res.send(200);
        }
      });
    }
  }); //  fs.readFile
};


/* helper functions */

function loadDir(folder, opt, callback) {
  fs.readdir(folder, function (err, filenames) {
    if (err) { callback(err); return; }
    filterMap(filenames, opt, function (filename, done) {
      fs.stat(path.join(folder, filename), function (err, stat) {
        if (err) { done(err); return; }
        if (stat.isFile()) {
          fs.readFile(path.join(folder, filename), done);
          return;
        }
        done();
      });
    }, callback); // filterMap
  });
};


function filterMap(array, opt, filter, callback) {
  var counter = array.length;
  var new_array = [];
  array.forEach(function (item, index) {
    filter(item, function (err, result) {
      if (err) { callback(err); return; }
      var json = JSON.parse(result);
      if (opt == null) {
        new_array[index] = json;
      }
      counter--;
      if (counter === 0) {
        new_array.length = array.length;
        callback(null, new_array.filter(function (item) {
          return typeof item !== 'undefined';
        }));
      }
    });
  });
};

function saveFile(filepath, id, filedata, ext, callback) {
  fs.writeFile(filepath + "/" + id + "." + ext,  filedata, function(err){
      if (err) {
        callback(err);
      } else {
        callback(null, filedata);
      }
  }); 
};

function getMaxFilename(filenames, callback) {
  var max = 1;
  var ii = 1;
  var len = filenames.length;
  filenames.forEach(function (filename) {
    max = Math.max(max, parseInt(filename, 10));
    if (len == ii) {
      callback(null, max+1);
    }
    ii++;
  });
};

function removeDir(folder, opt, callback) {
  fs.readdir(folder, function (err, filenames) {
    if (err) {
      callback(err);
    } else {
      removeData(filenames, function(filename, done) {
        fs.unlink(path.join(folder, filename), function(err) {
          if (err) {
            console.log(err);
            done(err);
            return;
          }
          done();
        });
      }, callback);
    }
  });
};

function removeData(array, filter, callback) {
  var counter = array.length;
  array.forEach(function (item, index) {
    filter(item, function(err) {
      if (err) {
        callback(err);
        return;
      }
      callback();
    });
  });
};

module.exports = RESTHelper;
